### PR TITLE
New audit.py module to replace nova (hubble.py)

### PR DESCRIPTION
**Pull request by basepi**
The new audit module shares a lot with the previous generation (nova). I cut out a few things that I don't think were getting much use (like controlled checks and compliance % calculations). If we decide we have uses for them, they would be pretty trivially easy to add to the new module.

I also added a couple of new features, `target` and `version`. Check out the docs in audit.py to get a feel for how they work, and keep me posted if you have any questions.

So far I have only ported a single module to the new system, grep.py. As you can see, the actual audit modules themselves are *vastly* simplified compared to nova. Porting the rest of them should be pretty trivial.

***

Two other missing pieces:

* Code to convert the old profiles to the new format. This code will be module-specific (grep checks will need different logic thank pkg checks, for example). I'm hopeful I can write something for grep next week.
* New audit returner. Honestly this should be pretty easy to whip up so I figured spending time on it in my remaining days was low priority. I think @Tenebriso or @jettero could whip this up in no time.

***

You may notice the lack of `nova_loader.py`. For this initial pass I decided to try it out without dynamic loading. Thus the manual process of registering audit functions in the global `audit_modules` dictionary. This has some drawbacks, like the lack of `__salt__` and `__grains__` in the underlying modules. However, without having to dynamically load the yaml files, I could probably follow the pattern I used in FDG to load these modules dynamically with an unmodified loader. I may do that this afternoon, but wanted to get this up for review now.

Edit: Turns out adding the loader was super easy. See new commit for the difference.

### CODE REVIEW

File: hubblestack/extmods/audit/grep.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    :param match_output_multiline:
+        See module-level documentation
+    :return:
+        Returns a tuple (success, {'grep_output': output}) where ``success``
+        is True or False based on the success of the check, and ``output`` is
+        the output of the ``grep`` command, for documentation purposes.
+    """
+    if not os.path.isfile(path):
+        if success_on_file_missing:
+            return True, {}
```
Comment (mew1033) [inline hubblestack/extmods/audit/grep.py]: Don't you need an

    else:
        return False, {}
I ran the salt-call cmd.run like it is in the code (I think), and got this:

    [root@hostn ~]# salt-call cmd.run 'grep hi /no/file' python_shell=False ignore_retcode=True
    local:
        grep: /no/file: No such file or directory
    [root@hostn ~]#

**Comment by mew1033** [comment]
Would a check like this still be possible?

    limit_su_command_access:
      data:
        'Ubuntu-18.04,Debian-8':
          - /etc/pam.d/su:
              pattern: pam_wheel.so
              match_output: ^auth\s+required\s+pam_wheel\.so\s+use_uid
              match_output_regex: true
              tag: CIS-su-command
          - /etc/group:
              pattern: sudo
              tag: CIS-su-command
        '*CoreOS*':
          - /etc/pam.d/su:
              pattern: pam_wheel.so
              match_output: ^auth\s+required\s+pam_wheel\.so\s+use_uid
              match_output_regex: true
              tag: CIS-su-command
        "*":
          - /etc/pam.d/su:
              pattern: pam_wheel.so
              match_output: ^auth\s+required\s+pam_wheel\.so\s+use_uid
              match_output_regex: true
              tag: CIS-su-command
          - /etc/group:
              pattern: wheel
              tag: CIS-su-command
      description: Restrict access to the su command

(This is the old format)
The check has multiple implementations based on OS. Or would the way to do this now be to just have multiple checks with the same id and different targets?

**Comment by mew1033** [comment]
Or a check like this? Again, in the old format:

    sshd_idle_timeout:
      data:
        '*':
          - /etc/ssh/sshd_config:
              # regex for 300 or less
              pattern: '"^\s*ClientAliveInterval\s+0*(([1-9][0-9]?)|((1|2)[0-9]{2})|(300))\s*(#.*)?$"'
              grep_args:
                - '-E'
              tag: CIS-ssh-timeout
          - /etc/ssh/sshd_config:
              pattern: '"^\s*ClientAliveCountMax\s+0*[0-3]\s*(#.*)?$"'
              grep_args:
                - '-E'
              tag: CIS-ssh-timeout
      description: Ensure SSH Idle Timeout Interval is configured

This requires that each of the things in '*' list pass before the check passes.

File: hubblestack/extmods/modules/audit.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        elif version:
+            version_match = _version_cmp(version)
+            if not version_match:
+                LOG.debug('Skipping audit {0} due to version {1} not matching '
+                          'version requirements {2}'
+                          .format(audit_id, __grains__['hubble_version'], version))
+                continue
+
+        # Process tags via globbing
+        if not fnmatch.fnmatch(tag, tags):
```
[the remark this answers is not in the document]
Comment (jettero) [reply hubblestack/extmods/modules/audit.py]: It'd be a pain to not have the fancy matching when you need it and have to wait a year to get it. Normally fnmatch is plenty (I've never used FDG, so maybe not relevant) but I've needed something more in targeting occasionally.

**Comment by jettero** [comment]
> Turns out adding the loader was super easy

_barely an inconvenience_?

File: hubblestack/extmods/modules/audit.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+            LOG.debug('Skipping audit {0} due to tag {1} not matching tags {2}'
+                      .format(audit_id, tag, tags))
+            continue
+
+        # Process labels
```
[the remark this answers is not in the document]
Comment (mew1033) [reply hubblestack/extmods/modules/audit.py]: Yup. Missed that. Thanks!

File: hubblestack/extmods/modules/audit.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+            success, data_dict = __audit__[module](*args, **kwargs)
+        except Exception as e:
+            LOG.error('Audit {0} from file {1} failed with exception {2}'
+                      .format(audit_id, audit_file, e))
+            data['reason'] = 'exception'
+            data['exception'] = str(e)
+            ret['Failure'].append({audit_id: data})
+            continue
+
+        if data_dict and isinstance(data_dict, dict):
```
Comment (mew1033) [inline hubblestack/extmods/modules/audit.py]: What if data_dict is not a dict

File: hubblestack/extmods/audit/grep.py (hunk 1 above)
Comment (basepi) [reply hubblestack/extmods/audit/grep.py]: Yes, I shall fix this.

File: hubblestack/extmods/modules/audit.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (basepi) [reply hubblestack/extmods/modules/audit.py]: As discussed this morning in our meeting, we don't use this feature internally and you were conflating your hatred of fnmatch with the old osfinger matching in nova. I think this is fine.

File: hubblestack/extmods/modules/audit.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        if not __salt__['match.compound'](target):
+            LOG.debug('Skipping audit {0} due to target mismatch: {1}'.format(target))
+            continue
+
+        args = data.get('args', [])
+        kwargs = data.get('kwargs', {})
+
+        # Run the audit
+        try:
+            success, data_dict = __audit__[module](*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (basepi) [reply hubblestack/extmods/modules/audit.py]: Tracked in https://jira.corp.adobe.com/browse/CSTTEAM-45153

File: hubblestack/extmods/modules/audit.py (hunk 4 above)
Comment (basepi) [reply hubblestack/extmods/modules/audit.py]: Fixed.

File: hubblestack/extmods/modules/audit.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+from __future__ import absolute_import
+
+import fnmatch
+import logging
+import os
+import salt.loader
+import salt.utils
+import yaml
+
+import hubblestack.extmods.audit.grep
```
[the remark this answers is not in the document]
Comment (basepi) [reply hubblestack/extmods/modules/audit.py]: Done.

File: hubblestack/extmods/audit/grep.py (hunk 1 above)
Comment (basepi) [reply hubblestack/extmods/audit/grep.py]: Done.

**Comment by basepi** [comment]
I put in the easy changes. For @mew1033's questions about those compound checks, they would work with some changes. For the first set of checks, they would turn into 5 disparate checks with the same tag and different IDs. Then, in addition to changing the target strings to use `G@osfinger:`, you would have to also change the `'*'` to be `and not G@osfinger:*CoreOS* and not G@osfinger:'Ubuntu-18.04 <etc>`.

Note that you could maybe make these a little bit easier since you can now target against all grains, but in the old nova it would find the first match and skip the rest, but in the new audit each target will be checked independently, so you have to do those `and not` shenanigans.

We could make checks aware of other checks but we'd need to hash out how that would be expressed in the data.

**Comment by basepi** [comment]
Also added the `Skipped` key that we talked about in returns. It only includes checks skipped via version checking

**Comment by praksinha** [comment]
@basepi Looks good to me. Thanks for kick-starting the nova re-write :)

**Pull request merged**